Assets/__Scripts/Beatmap/Animations/PointDefinition.cs

```diff
@@ -0,0 +1,285 @@
+                    next = m;
+                }
+            }
+        }
+
+        public int CompareTo(PointDefinition<T> other)
+        {
+            // TODO: might be able to cheese previous/next with this
+            return StartTime.CompareTo(other.StartTime);
+        }
+
+        public class PointData : IComparable<PointData>
+        {
+            public PointData(Parser parser, JSONArray data, float tbegin = 0, float tend = 0)
+            {
+                Value = parser(data, out int i);
+                var len = data.Count;
+                if (len > i)
+                {
+                    // Track or Path animation
+                    Time = (tend == 0)
+                        ? data[i++].AsFloat
+                        : Mathf.LerpUnclamped(tbegin, tend, data[i++]);
+                }
+                else
+                {
+                    // WTF Jevk
+                    Time = 0;
+                }
+                Easing = global::Easing.Linear;
+                Lerp = PointDataInterpolators.LinearLerp<T>;
+
+                for (; i < len; ++i)
+                {
+                    string str = data[i];
+                    if (str[0] == 'e')
+                    {
+                        Easing = global::Easing.Named(str);
+                    }
+                    if (str == "splineCatmullRom")
+                    {
+                        Lerp = PointDataInterpolators.CatmullRomLerp<T>;
+                    }
+                    if (str == "lerpHSV")
+                    {
+                        Lerp = PointDataInterpolators.HSVLerp<T>;
+                    }
+                }
+            }
+
+            public T Value { get; }
+            public float Time { get; }
+            public Func<float, float> Easing { get; }
+            public InterpolationHandler Lerp { get; }
+
+            public int CompareTo(PointData other)
+            {
+                return Time.CompareTo(other.Time);
+            }
+        }
+    }
+
+    public class PointDataInterpolators
+    {
+        public static T LinearLerp<T>(PointDefinition<T>.PointData[] points, int prev, int next, float time) where T : struct
+        {
+            return LinearLerpFunc<T>()(points[prev].Value, points[next].Value, time);
+        }
+
+        public delegate T LerpFunc<T>(T prev, T next, float time);
+        static LerpFunc<float> LinearFloat = new LerpFunc<float>(Mathf.LerpUnclamped);
+        static LerpFunc<Color> LinearColor = new LerpFunc<Color>(Color.LerpUnclamped);
+        static LerpFunc<Vector3> LinearVector3 = new LerpFunc<Vector3>(Vector3.LerpUnclamped);
+        static LerpFunc<Quaternion> LinearQuaternion = new LerpFunc<Quaternion>(Quaternion.SlerpUnclamped);
+
+        public static LerpFunc<T> LinearLerpFunc<T>() where T : struct
+        {
+            // I hate C#
+            return typeof(T) switch
+            {
+                var n when n == typeof(float) => (LerpFunc<T>)(object)(LinearFloat),
```

Comment:
what are the performance costs of boxing/unboxing like this? if its substantial, we may need to rethink this approach.

Reply:
Found a different, maybe cleaner way to do this in 26f57e2.

Performance impact of any of these are negligible, old:
![Screenshot from 2023-09-17 10:05:34](https://github.com/Caeden117/ChroMapper/assets/74876786/3d81e37c-93a0-4c83-b6f6-eb2835304793)
vs new:
![Screenshot from 2023-09-17 10:08:34](https://github.com/Caeden117/ChroMapper/assets/74876786/db6ca06b-5218-4e98-ab16-76980413e5f3)

The actual lerp function call only accounts for 0.05ms of both.